### DISCUSSION TITLE
Unnecessary parentheses?

### DISCUSSION THREAD

**Discussion by kozaka-tv** [Th0rstenf/Rocksniffer-to-Streamer.bot-extension#114] [General]
In some of my commit (18fa4a8), I removed some parentheses, as they are not needed at all. 
Then some came back with commit e0cd893 with the comment: "Reintroduced a few parentheses, as they make it more explicit and don't rely on potentially compiler implementation specific order of evaluation"

So I thought maybe we could discuss this, and learn from it. As I am a very beginner in C#, need to read always some documentations in case, there is a diff between Java and C#. But, I trust the Rider IDE, that what there is a warning/note, is probably something good to change (but not always that is the case, but many-many times).

![image](https://user-images.githubusercontent.com/4964462/231879242-b689cbd4-df3f-45fc-ae21-f74d360ae32a.png)

I searched on the internet what are the rules, and found out, that in C# there is a strict left to right. If compiler thinks, there is no side effect, it could be, that the evaluation is in different order than as we think, but this should not change the result.

> In the C# language, the order of evaluation is spelled out more explicitly. [The order of evaluation for operators is left to right](http://msdn2.microsoft.com/en-us/library/Aa691322). if you write f() + g() in C#, the language guarantees that f() will be evaluated first. The example in the linked-to page is even clearer. The expression F(i) + G(i++) * H(i) is evaluated as if it were written like this:

> The side effects of each part of the expression take effect in left-to-right order. Even [the order of evaluation of function arguments is strictly left-to-right](http://msdn2.microsoft.com/en-us/library/Aa691335).

> Note that [the compiler has permission to evaluate the operands in a different order if it can prove that the alternate order of evaluation has the same effect as the original one](http://msdn2.microsoft.com/en-us/library/Aa691105) (in the absence of asynchronous exceptions).

https://devblogs.microsoft.com/oldnewthing/20070814-00/?p=25593
https://stackoverflow.com/questions/4644328/behaviour-and-order-of-evaluation-in-c-sharp
https://stackoverflow.com/questions/1215236/c-order-of-function-evaluation-vs-c

**Comment by Th0rstenf** [comment]
You've searched up the wrong things, have a look at this:
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/#operator-precedence

Operators evaluation order is depending on their precedence.
so an evaluation:
```
if (CONDITION && Number < 0)
```
_should_ be evaluated as 
```
CONDITION && ( Number < 0)
```
_Note that left to right would already be horrible here_
As defined by the language, that's why your IDE gives you a note (shouldn't be a warning) that it is redundant.
_**HOWEVER**_ There have been so many deviations from language standard in different compiler implementations in the past, that this was not always the case!
Therefore static code analysis would actually throw a message for this line without the parentheses.
And as that is what i'm using at work, i'm used to doing it like this. :shrug:
